File: ios/Classes/models/action_sheet/FCPActionSheetTemplate.swift
//
//  FCPActionSheetTemplate.swift
//  flutter_carplay
//
//  Created by Oğuzhan Atalay on 25.08.2021.
//

import CarPlay

@available(iOS 14.0, *)
class FCPActionSheetTemplate {
  private(set) var _super: CPActionSheetTemplate?
  private(set) var elementId: String
  private var title: String?
  private var message: String?
  private var tabTitle: String?
  private var systemIcon: String?
  private var showsTabBadge: Bool = false
  private var actions: [CPAlertAction]
  private var objcActions: [FCPAlertAction]

  init(obj: [String: Any]) {
    self.elementId = obj["_elementId"] as! String
    self.title = obj["title"] as? String
    self.message = obj["message"] as? String
    self.tabTitle = obj["tabTitle"] as? String
    self.systemIcon = obj["systemIcon"] as? String
    self.showsTabBadge = obj["showsTabBadge"] as! Bool
    self.objcActions = (obj["actions"] as! [[String: Any]]).map {
      FCPAlertAction(obj: $0, type: FCPAlertActionTypes.ACTION_SHEET)
    }
    self.actions = self.objcActions.map {
      $0.get
    }
  }

  var get: CPActionSheetTemplate {
    let actionSheetTemplate = CPActionSheetTemplate.init(
      title: self.title, message: self.message, actions: self.actions)
    actionSheetTemplate.tabTitle = tabTitle
    actionSheetTemplate.showsTabBadge = showsTabBadge
    if let systemIcon = systemIcon {
      actionSheetTemplate.tabImage = UIImage(systemName: systemIcon)
    }

    actionSheetTemplate.elementId = self.elementId
    self._super = actionSheetTemplate
    return actionSheetTemplate
  }

  public func update(with: FCPActionSheetTemplate) {
  }
}

@available(iOS 14.0, *)
extension FCPActionSheetTemplate: FCPPresentTemplate {}

File: android/src/main/kotlin/com/oguzhnatly/flutter_carplay/models/action_sheet/FCPActionSheetTemplate.kt
package com.oguzhnatly.flutter_carplay.models.action_sheet

import androidx.car.app.model.Action
import androidx.car.app.model.ActionStrip
import androidx.car.app.model.LongMessageTemplate
import com.oguzhnatly.flutter_carplay.CPActionSheetTemplate
import com.oguzhnatly.flutter_carplay.FCPPresentTemplate
import com.oguzhnatly.flutter_carplay.models.alert.FCPAlertAction

/**
 * A wrapper class for CPActionSheetTemplate with additional functionality.
 *
 * @param obj A map containing information about the action sheet template.
 */
class FCPActionSheetTemplate(obj: Map<String, Any>) : FCPPresentTemplate() {

    /// The underlying CPActionSheetTemplate instance.
    private lateinit var _super: CPActionSheetTemplate

    /// The title of the action sheet template (optional).
    private var title: String?

    /// The message of the action sheet template (optional).
    private var message: String?

    /// An array of CPAlertAction instances associated with the action sheet template.
    private var actions: List<Action>

    /// An array of FCPAlertAction instances associated with the action sheet template.
    private var objcActions: List<FCPAlertAction>

    init {
        val elementIdValue = obj["_elementId"] as? String
        assert(elementIdValue != null) {
            "Missing required key: _elementId"
        }
        elementId = elementIdValue!!

        title = obj["title"] as? String
        message = obj["message"] as? String

        objcActions = (obj["actions"] as? List<Map<String, Any>>)?.map {
            FCPAlertAction(it)
        } ?: emptyList()
        actions = objcActions.map { it.getTemplate() }
    }

    /** Returns the underlying CPActionSheetTemplate instance configured with the specified properties. */
    override fun getTemplate(): CPActionSheetTemplate {
        val actionSheetTemplate =
            LongMessageTemplate.Builder(message.toString()).setTitle(title.toString())
        objcActions.forEach {
            when (it.style) {
                CPAlertActionStyle.destructive, CPAlertActionStyle.cancel -> {
                    actionSheetTemplate.addAction(it.getTemplate())
                }

                else -> {
                    val actionStripBuilder = ActionStrip.Builder()
                    actionStripBuilder.addAction(it.getTemplate())
                    actionSheetTemplate.setActionStrip(actionStripBuilder.build())
                }
            }
        }
        _super = actionSheetTemplate.build()
        return _super
    }

}
